Issue #3907: SummaryJavadoc no violation on javadoc where '.' is used…

sagar-shah94 opened the pull request:
Issue: #3907

sagar-shah94 commented:
@rnveach @romani 
I am using english message in every message.properties fiile.
I dont know other languages how can i set messages for those languages?

romani commented:
> I dont know other languages how can i set messages for those languages?

please use google translator. If eventually native people find a mistake they will send PR.

sagar-shah94 commented:
@rnveach @romani 
These errors are coming in CI because in some files are not written properly.
for example -https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L122
in above java doc description is not given .
Please verify errors and check

rnveach commented on src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule72thesummaryfragment/InputCorrectSummaryJavaDocCheck.java:
```diff
@@ -3,23 +3,18 @@
 /**
  * Some Javadoc A {@code Foo} is a simple Javadoc.
  */
-class InputCorrectSummaryJavaDocCheck {
```
Please revert all superfluous changes to this input. It is making it hard to to understand what was changed.
Any new cases you have to add, please add them to the end or in a separate file.

rnveach commented on src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule72thesummaryfragment/InputIncorrectSummaryJavaDocCheck.java:
```diff
@@ -5,17 +5,17 @@
  * is a simple Javadoc. Some javadoc.
  */
 class InputIncorrectSummaryJavaDocCheck {
-    
```
Same here.

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/InputIncorrectSummaryJavaDoc.java:
```diff
@@ -5,17 +5,17 @@
  * is a simple Javadoc. Some javadoc.
  */
 class InputIncorrectSummaryJavaDoc {
-    
```
And here.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -82,6 +89,9 @@
     private static final String PERIOD = ".";
 
+    /** inherit doc literal. */
+    private static final String INHERIT_DOC = "{@inheritDoc}";
```
Shouldn't we ignore `inheritDoc` as it will be covered by PR #4047 ?

sagar-shah94 replied:
Yeah i was working on that.
https://github.com/checkstyle/checkstyle/pull/3926
Above is pull request for that, all asked changes have been done.
But until it gets merged i had to handle that case that is why added it.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -134,19 +144,71 @@ public void setPeriod(String period) {
-        if (endOfSentence == -1) {
-            if (!firstSentence.trim().startsWith("{@inheritDoc}")) {
+        boolean temp = true;
+        if (period.equals(PERIOD)) {
```
Why are we checking these 2?
If user supplied custom `period` we should be using that as the period and nothing else.

sagar-shah94 commented:
@rnveach made changes,please verify.

sagar-shah94 commented:
@rnveach these CI are failing because we have not written proper javadoc at some places
for example -https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L122
in above java doc description is not given .
Please verify errors and check

rnveach commented:
@sagar-shah94 

> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser.java:459: First sentence of Javadoc is incomplete (period is missing) or not present. [SummaryJavadoc]

This looks incorrect. `Logs parser errors in Checkstyle manner.` is the first sentence and ends with a period.

> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java:392: Summary javadoc is missing. [SummaryJavadoc]

This looks false. `Custom ResourceBundle.Control implementation which allows explicitly read the properties files as UTF-8.` is the first sentence and ends with a period.

> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/api/SeverityLevel.java:24: Summary javadoc is missing. [SummaryJavadoc]

Same as previous. `Severity level for a check violation.` is first sentence.

> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java:122: Summary javadoc is missing. [SummaryJavadoc]

This case looks valid and are a few others being reported.

romani commented:
Changes for checkstyle code for missed first lines should be in separate commit of the same PR ( two commits in this PR is expected)

rnveach commented:
@sagar-shah94 ping
Since we are out of proposal deadline, do you think you can finish this?
Please rebase off latest master as there are conflicts.

sagar-shah94 commented:
please verify - https://github.com/checkstyle/checkstyle/pull/4084
currently working on -https://github.com/checkstyle/checkstyle/pull/3998
once i am done with above two pull requests , i will start working on this issue.

MEZk commented:
@sagar-shah94 
Please rebase your changes onto the latest master.

rnveach commented:
@sagar-shah94 ping
This is now your oldest PR.

sagar-shah94 commented:
Ok will start working on this.
@rnveach

sagar-shah94 commented:
@rnveach which should i work on first? this PR or https://github.com/checkstyle/checkstyle/pull/4174

because another one is subpart of this PR.

rnveach commented:
@sagar-shah94 I don't think either should conflict with each other as one is only for a custom period (non-default check) and the other is for identifying the end of a sentence using said period (default check).
If you have to choose, I would say this one, but you could work on both until you see if one of your implementations does indeed rely on the other.

rnveach commented:
@sagar-shah94 ping

sagar-shah94 commented:
@rnveach currently working on this issue but i am stuck in one place.

sagar-shah94 commented:
I am confused in one case.
Which one of below is correct?
<p>comment.</p>
<p>comment</p>.
@rnveach

rnveach commented:
> Which one of below is correct?

We should follow what javadoc util does. Check and violation is based on it pulling in the first sentence and placing it in the summary for the HTML report, but there still could be some scenarios we don't agree with.
Ideally `<p>comment.</p>` is the correct way.

This is the results of the javadoc util.

![cs_period](https://cloud.githubusercontent.com/assets/5427943/25046313/c45fbf6e-20fe-11e7-83be-2af4666395ab.png)

It doesn't recognize the period outside the `<p>` tag.

sagar-shah94 commented:
I was stuck that is why it is taking much time.Thanks for clarifying my doubts.
@rnveach will try to fix this issue as soon as possible .

sagar-shah94 commented:
@rnveach  please verify. CI are failing because javadoc is not properly written at some places.if anything is wrong let me know .if nothing is wrong , i will have to fix cyclometric complexity because it is 17 now.Once you approve code is working properly i will fix cyclometric complexity

sagar-shah94 commented:
@rnveach https://github.com/checkstyle/checkstyle/issues/2991 issue will be fixed along with this issue.

romani commented:
@sagar-shah94 ,

> I was stuck that is why it is taking much time.

Please feel free to move `<p>comment.</p>` case to separate issue. For our validation now, ignoring of all tags in search for first `.` is more than enough. So would recommend to treat
```
<p>comment.</p>
<p>comment</p>.
```
as both correct cases.

> CI.......Checkstyle failed: Got 124 errors and 0 warnings.

please provide fix for our bad javadocs in separate commit of this PR. I recommend place it as first to let you make extra changes in commit for logic with less manual work.

rnveach commented:
> Checkstyle failed: Got 124 errors and 0 warnings.

> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyOption.java:63: First sentence of Javadoc is incomplete (period is missing) or not present. [SummaryJavadoc]

First sentence is `Represents the policy that the brace must always be on a new line.` and it has a period.

> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java:142: First sentence of Javadoc is incomplete (period is missing) or not present. [SummaryJavadoc]

First sentence is `Determines if ast is in the body of a flow control statement without braces.` and it has a period.

> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:441: First sentence of Javadoc is incomplete (period is missing) or not present. [SummaryJavadoc]

First sentence is `Checks if the given method name matches the regex.` and it has a period.

Like romani said, I think you should look at these errors yourself and start fixing cases.
It may help you find ones that are surely wrong too as the ones I mentioned above can be clearly seen to have a period in them.

sagar-shah94 commented:
@rnveach @romani 
```
   /**
     * Represents the policy that the brace must always be on a new line. For
     * example:
     * <pre>
     * if (condition)
     * {
     *     ...
     * </pre>
     */
```
I think this is invalid because "**Represents the policy that the brace must always be on a new line. For example:**" i am considering whole as a sentence and period is not at the end.


Because in romani's example
```
    /**
     * JAXB 1.0 only default validation event handler
     */
```
**JAXB 1.0 only default validation event handler** is considered as a sentence .not just **JAXB 1.**
Please correct me if i am wrong.

rnveach commented:
> i am considering whole as a sentence and period is not at the end.

Please read violation message check is printing for these lines and the description of the check at http://checkstyle.sourceforge.net/config_javadoc.html#SummaryJavadoc.
>  First sentence of Javadoc is incomplete (period is missing) or not present.
> Checks that Javadoc summary sentence ...

Both are saying 'first sentence', not 'last' sentence. This is why I pointed these lines out as false positives.
This issue is only about fixing cases where there is a period located in the javadoc, but it should not be considered the end of the first sentence.

> `JAXB 1.0 only default validation event handler` is considered as a sentence .not just JAXB 1. Please correct me if i am wrong.

You are correct here. `1.0` is not a end of sentence period, just a version number. This example is also the first sentence of the javadoc.

sagar-shah94 commented:
@rnveach please verify one more time and let me know if it is correct.
I checked error by myself now i think it is correct.
Let me know if i have misunderstood anything about this issue.

romani commented:
@sagar-shah94 , please do not wait and fix CIs problems, as you start fixing them you will see all false-positives yourself. 
Please do not forget to put into UTs all cases that you stumble during resolving issues in our code.

We will continue review when CI become green

sagar-shah94 commented:
@rnveach @romani 
Can you please provide me content for javadoc wherever javadoc is missing and error is being generated.
Whereever period was missing at the end of sentence, i have added and fixed that error.
But to fix missing javadoc, I need content.

Vladlis commented:
@sagar-shah94
You should not modify existing input files.

sagar-shah94 commented:
@Vladlis 
When summary java doc is not there ,we are showing "summary java doc is missing",that is why changed files.

rnveach commented:
@sagar-shah94 CI needs to be green.

> Can you please provide me content for javadoc wherever javadoc is missing 

Do the best you can for completely missing javadocs. If we think it is missing somethings, we will help you write a better one.

sagar-shah94 commented:
@rnveach ok sir .will try to fix as soon as possible

sagar-shah94 commented:
@romani @rnveach please verify and suggest changes if needed.

rnveach commented on src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule72thesummaryfragment/InputCorrectSummaryJavaDocCheck.java:
```diff
@@ -9,17 +9,18 @@
      */
     public static final byte NUL = 0;
-    
+
```
Please undo superfluous changes in inputs like this and others.

sagar-shah94 replied:
Done.

rnveach replied:
This is not done.

rnveach commented on src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule72thesummaryfragment/InputCorrectSummaryJavaDocCheck.java:
```diff
@@ -29,63 +30,66 @@ void foo5() {}
-     * <a href="mailto:vlad@htmlbook.ru"/> 
+    /**
+     * This is valid.
+     * <a href="mailto:vlad@htmlbook.ru"/>.
```
If you fix a 'bad' javadoc in a good 'input' file please make sure to make a copy of it in the bad 'input' file if it isn't already.

sagar-shah94 replied:
Done.

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/InputIncorrectSummaryJavaDoc.java:
```diff
@@ -5,17 +5,17 @@
+
     /**
-     * As of JDK 1.1, replaced by {@link #setBounds(int,int,int,int)}
+     * As of JDK 1.1, replaced by {@link #setBounds(int,int,int,int)}.
```
don't fix 'bad' javadocs in a 'bad' input file.
We expect violations here to make sure your logic is working correctly.

sagar-shah94 replied:
Done.

romani commented:
@sagar-shah94 , ping

sagar-shah94 commented:
@romani sir 
I have report to submit in college so i was busy with that.
I am currently working on that report only. that has to be submitted on 29th april.once it is done from very next day i will start working on this issue.

sagar-shah94 commented:
@rnveach in my local mvn clean verify showed BUILD SUCCESS , here CI are failing .Should i rebase 
 and push again?

romani commented:
@sagar-shah94 , YES please do.
 
it is always good to be based on latest code (rebased on main master).
in this case you can compare build status on main master and your issue/topic branch.

sagar-shah94 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -19,7 +19,6 @@
 
 package com.puppycrawl.tools.checkstyle.checks.javadoc;
 
-import java.util.Arrays;
```
Done.

sagar-shah94 commented:
@romani @rnveach suggested changes are done. please verify and suggest changes if needed.

romani commented:
@sagar-shah94, please move second commit to separate PR to let us merge it quicker and keep less changes on your side.

first commit please split into two not related to fix formatting and functional changes. Formatting  could go to separate PR to be merged quicker. This will help to keep functional changes in minimal state and speed up functional fix acceptance.

romani commented:
@sagar-shah94 , ping. Please do separate PR as I suggested to simplify this PR.

sagar-shah94 commented:
@romani i have already created separate PR .please verify.
https://github.com/checkstyle/checkstyle/pull/4413

romani commented:
@sagar-shah94 , please close this PR if all is moved to new PR

rnveach commented:
@romani all is not moved to new PR.
We decided to split this into 2 commits, changes to check and violation fixes.
Only violation fixes were moved to second PR.

sagar-shah94 commented:
@rnveach i think only violations in comments need to be moved that i have already done.

romani commented:
@sagar-shah94 ,

I do not see in github your latest rebased code, there should be commit reference after my/this comment in this PR.
please push all to remote to let me see your latest code and CI feedback on it.

rnveach commented:
@sagar-shah94 If you make changes as we have requested, please notify us. We don't get notifications on new commits.
Are you ready for us to continue reviewing?

sagar-shah94 commented:
@rnveach yes.sorry forgot to notify u.

sagar-shah94 commented:
@rnveach i have to remove unnecessary spaces from code.

rnveach commented:
@sagar-shah94 As you make new changes, please rebase on latest master.

sagar-shah94 commented:
@rnveach next time will take care of this.will rebase every time when i will push changes

romani commented:
@sagar-shah94 , I we mentioned before - move all spaces changes in separate PR. They should not be present in this PR.

romani commented:
@sagar-shah94 , please rebase and lets finish this PR, you are almost done.

sagar-shah94 commented:
ok i will finish this PR today

sagar-shah94 commented:
@romani removed superfluous spaces.please verify

sagar-shah94 commented:
@romani @rnveach  
How can i check if string is empty or not?
i used `isEmpty()` function but when i ran mvn clean verify it failed?
now CI is failing here.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -159,35 +155,101 @@ public void visitJavadocToken(DetailNode ast) {
+        return child.getType() == JavadocTokenTypes.NEWLINE
+                && !result.toString().trim().equals("")
+                && String.valueOf(result.toString().trim()
+                .charAt(result.toString().trim().length() - 1)).equals(period);
```
please intend +4 spaces

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -159,35 +155,101 @@ public void visitJavadocToken(DetailNode ast) {
+        else if (!periodFound
+                    && !period.isEmpty()
+                        && !str.contains(period)
+                            && !str.equals(INHERIT_DOC)) {
```
All "&&" should have the same indentation

sagar-shah94 replied:
Done.

sagar-shah94 commented:
@romani suggested changes are made, please verify and let me know if more changes are needed.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -159,35 +155,101 @@ public void visitJavadocToken(DetailNode ast) {
+     * @param child child node
+     * @return true if period is at end of first sentence.
+     */
+    private boolean isPeriodAtEnd(StringBuilder result, DetailNode child) {
```
argument should be String, `result.toString().trim()` should be done outside of this method.

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -159,35 +155,101 @@ public void visitJavadocToken(DetailNode ast) {
+     */
+    private void printWarningForViolation(DetailNode ast,
+                                          boolean periodFound,
+                                          StringBuilder result) {
```
argument should be String

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -159,35 +155,101 @@ public void visitJavadocToken(DetailNode ast) {
+     * @param periodFound true if period is found in sentence before end of line
+     * @param result javadoc string
+     */
+    private void printWarningForViolation(DetailNode ast,
```
`printWarningForViolation` -> `validate`

sagar-shah94 replied:
Done.

sagar-shah94 commented:
![image](https://user-images.githubusercontent.com/6566854/27518434-bcfdf3d4-59fd-11e7-8d2e-c161957710ee.png)
This error occured when i tried to generate report.
@romani @rnveach

rnveach commented:
@sagar-shah94 I don't know the error, but it is coming from the `hg` command.
For now, please comment out the `openjdk` projects and run regression on the rest.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -159,35 +156,101 @@ public void visitJavadocToken(DetailNode ast) {
+     * @param result javadoc string
+     * @param detailNode javadoc tag node
+     */
+    private void getStringInsideTag(StringBuilder result, DetailNode detailNode) {
```
do not put StringBuilder as argument.
return String from this method.

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -159,35 +156,101 @@ public void visitJavadocToken(DetailNode ast) {
+    private void checkPeriodAtEnd(DetailNode ast) {
+        final Set<Integer> allowedTypes = new HashSet<>();
+        allowedTypes.add(JavadocTokenTypes.TEXT);
+        allowedTypes.add(JavadocTokenTypes.WS);
```
please move `allowedTypes` to class field, it is not going to be changed so has to be ImmutableSet.

sagar-shah94 replied:
@romani 
It is not allowing me to use ImmutableSet in code. it throws following error.
` Disallowed import - com.google.common.collect.ImmutableSet. [ImportControl]`

rnveach replied:
@sagar-shah94 Please just use `Collections.unmodifiableSet`.

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -145,12 +146,8 @@ public void setPeriod(String period) {
-            }
-        }
-        else {
+        checkPeriodAtEnd(ast);
```
method `visitJavadocToken` should contain all `log` statements. All methods should return results to make final decision at `visitJavadocToken`. Please refactor `checkPeriodAtEnd` method to return results to let you inline your `validate` method to `visitJavadocToken`

sagar-shah94 replied:
Done.

sagar-shah94 commented:
@romani will make changes soon

romani commented:
checkstyle 8.0 was released, please rebase on latest master. 
after each release it is mandatory to rebase to avoid CIs failures.

sagar-shah94 commented:
@romani @rnveach 
i am getting following error.how can i fix this?
`Line matches the illegal pattern '\s+$'. `

rnveach commented:
@sagar-shah94 Remove spaces from the end of the line. It should be only asking for that in production code, not the tests.

sagar-shah94 commented:
@romani @rnveach please verify.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -84,22 +92,22 @@
+    /** Inherit doc literal. */
+    private static final String INHERIT_DOC = "{@inheritDoc}";
+
+    /** Set of allowed javadoc tags in summary java doc. */
```
allowed Tokens

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -84,22 +92,22 @@
-                       JavadocTokenTypes.LEADING_ASTERISK,
-                       JavadocTokenTypes.EOF)
+    /** Inherit doc literal. */
+    private static final String INHERIT_DOC = "{@inheritDoc}";
```
it is already there - https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/JavadocTokenTypes.java#L639

you do not need your own definition of it.

sagar-shah94 replied:
i dont want to match token i want to match exact string that is why using this.

rnveach replied:
@sagar-shah94 Why don't you want to match against token? it is the most precise way to find a specific thing like the inherit doc literal. Match text is slower and not as accurate.
Some tokens like descriptions could have same text as inherit doc but not be true inherit docs.
We use parsers for a reason, you should rely on them more.

rnveach replied:
@romani @sagar-shah94 Are we going to fix this?

sagar-shah94 replied:
@rnveach can you please tell me how we can match against token?
i think we need to use absolute string to match. Let me know if i am wrong.

rnveach replied:
@sagar-shah94 One example in your current code is: https://github.com/checkstyle/checkstyle/pull/4047/files#diff-0bacc6c5fac536bdd9bde091d0bb53beR171

This will be part of new issue.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -84,22 +92,22 @@
-     */
+
+    /** Delimiter to separate line number from log message. */
+    private static final String DELIMITER = "#";
```
should be eliminated, see comment below

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -134,12 +142,14 @@ public void setPeriod(String period) {
     public void visitJavadocToken(DetailNode ast) {
```
logic is still over complicated.

you need to have visitJavadocToken method to be like:
```
String firstSentence = getFirstSentence(ast, period);
if (firstSentence.empty) {
   log(MSG_SUMMARY_JAVADOC_MISSING)
} else if (! isValid(firstSentence)) {
   log(MSG_SUMMARY_FIRST_SENTENCE)
}
else {
   if (containsForbiddenFragment(firstSentence)) { .....}
}
```

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -148,35 +158,106 @@ public void visitJavadocToken(DetailNode ast) {
+     * @param result javadoc string
+     * @return error string
+     */
+    private String validate(DetailNode ast,
```
please return string to `visitJavadocToken` method and do all validation of String that is first sentence there.

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -148,35 +158,106 @@ public void visitJavadocToken(DetailNode ast) {
-    private static boolean isOnlyInheritDoc(DetailNode ast) {
-        boolean extraTextFound = false;
-        boolean containsInheritDoc = false;
+    private String checkPeriodAtEnd(DetailNode ast) {
```
all checks should be done in `visitJavadocToken` method, all you need is to have smart method that return String of first sentence from javadoc content.

sagar-shah94 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -148,35 +158,106 @@ public void visitJavadocToken(DetailNode ast) {
+    private String validate(DetailNode ast,
+                          boolean periodFound,
+                          String result) {
+        final String str = result.replace('*', ' ').trim();
```
> .replace('*', ' ')

that is incorrect, as you will miss all "*" in body of javadoc. If you do not need leading "*", you should not add them when you create result string.
Please add UT case on this and share a name of test method with us to recheck.

sagar-shah94 replied:
Done.

romani replied:
@sagar-shah94 , make a javadoc like 
```
/** method find '*' 
*    in string. Second sentence.
*/ 
``` . 
to have "*" not only as leading asterisks, but as part of text.
Check config of test should violate usage of "*" in first message.

sagar-shah94 commented:
@romani please verify

rnveach commented on src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule72thesummaryfragment/InputIncorrectSummaryJavaDocCheck.java:
```diff
@@ -85,32 +85,32 @@ void foo6() {}
-         * Some Javadoc.
-         */
-        public static final byte NUL = 0;
+            /**
```
These changes look superflorous. You just indented code over and I don't see a reason for it.
Please undo.

rnveach replied:
@sagar-shah94 I assume you are referring to this.
It looks like you indented this code over to the right 4 spaces.
I don't see a reason why we need to make this change, so I was asking you to undo the indent and move them to the left 4 spaces.

romani replied:
you did indentation of comment to the right, it is not required in your issue, please do not do this.
We expect now change in this lines.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -134,12 +139,16 @@ public void setPeriod(String period) {
-            if (!isOnlyInheritDoc(ast)) {
-                log(ast.getLineNumber(), MSG_SUMMARY_FIRST_SENTENCE);
-            }
+        final String summaryDoc = checkPeriodAtEnd(ast).trim();
```
> .trim()

should be done inside method that return Summary

>`
String firstSentence = getFirstSentence(ast);
...
String summaryDoc = checkPeriodAtEnd(ast)
`

in terms of Check, first sentence and summary is the same.
It is not clear why you calculte it by two different methods.
`checkPeriodAtEnd` should be renamed to `getSummarySentence` as minimum.

sagar-shah94 replied:
Done.

romani commented:
@sagar-shah94 , let me know if smth is not clear in our code review comments.
please rebase to resolve comments.

sagar-shah94 commented:
@romani i could not understand rnveach's last comment

romani commented:
> i could not understand rnveach's last comment

I extended the comment.

sagar-shah94 commented:
@romani @rnveach rebased and made suggested changes

sagar-shah94 commented:
CI are failing because summary java doc is missing in some of existing files.
https://ci.appveyor.com/project/Checkstyle/checkstyle/build/8621/job/5mpto1jplfcq8c1e#L3292

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:
```diff
@@ -134,12 +139,16 @@ public void setPeriod(String period) {
-            if (!isOnlyInheritDoc(ast)) {
-                log(ast.getLineNumber(), MSG_SUMMARY_FIRST_SENTENCE);
-            }
+        final String summaryDoc = getSummarySentence(ast);
```
@sagar-shah94 , please explain why we have two method `getFirstSentence` and `getSummarySentence` ?
This methods should do absolutely the same as first sentence is summary sentence, please use only one.
If smth is missed in one of them, please merge them.

sagar-shah94 replied:
I will not be able to merge both methods into one because in getSummarySentence i am trying to fetch details upto first period, and in getFirstSentence we will be getting whole first sentence.I have been trying to merge these 2 methods but is not working because purpose of both methods are totally different that is what the reason why i had to introduce my own method.

romani commented:
@sagar-shah94 , please find time to finish this PR, there are more and more bugs are opening on this Check and do some fixes in it while you are in progress and almost done have little sense.

sagar-shah94 commented:
@romani i was confused that do i have to  fix bad summary java doc that caused CI to fail in same PR or create a separater for that?

romani commented:
such failures will be fixed by #4771 , please rebase after it is merged.

sagar-shah94 commented:
@romani I have addressed why we can not merge both methods , please reply to that in last change you suggested.

romani commented:
@sagar-shah94 , please rebase to fix CI problems.

please run regression testing by diff.groovy from https://github.com/checkstyle/contribution/tree/master/checkstyle-tester , please review report your self and share with us, all instructions are README

sagar-shah94 commented:
diff report=sagar-shah94.github.io

sagar-shah94 commented:
@romani thanks for approving

romani commented:
@rnveach , please do final review.

I am ok with report and changes. Two issues will be created on merge:
1) do investigation on two methods that suppose to return the same string (First sentence is the same as Summary sentence)
2) Messages: 
`First sentence of Javadoc is incomplete (period is missing) or not present.`
`Summary javadoc is missing.`
are overlapping. We need to investigate ability to remove `or not present`.

rnveach commented:
I am merging.

> Two issues will be created on merge:

We should also create an issue on `INHERIT_DOC` field string in class.
